### DISCUSSION TITLE
Implement for parameter finding

### DISCUSSION THREAD

**Discussion by kunxianhuang** [solvcon/modmesh#341] [Ideas]
An application estimates the parameters of the defined functions on some numerical data. We use maximum likelihood method as an estimation, and finding maximum/minimum is an optimization problem.

First, we have to decide the architecture of this application.
c++ in the bottom 
Python API

For the coding side, I have little experiences about programming architecture. 

Some objects need to complete as below
* Defined functions:
    * Gaussian function
    * Exponential function
    * Polynomial
    * Chebyshev polynomial
    * Poisson function
    * Crystal ball function:  The Crystal ball function is a Gaussian with a tail on the low side (or both side).
    * User defined function: it should be the same with scipy's user defined function
    * Stack up of above functions (or Call "ADD")
    * 
* First derivative vector of defined functions
* Hessian matrix of defined functions
* Constrains (parameters bound): for example $0<\mu<20.0$ and $0.5<\sigma<3.0$ for Gaussian function
* Log-likelihood value 
* Minimization methods
    * Iterator
    * Stop 
    * Linear programming for constrained minimization
* Plots

**Comment by yungyuc** [comment]
Thanks @kunxianhuang .  To get started, we may pick one or two functions and add prototype code.  What functions would you like to do first?

**Reply by yungyuc** [reply]
The issue has been quiet for long time.  I am turning this to a discussion conversation.  @kunxianhuang, do you have new ideas about this?

**Reply by kunxianhuang** [reply]
Sorry for quite long. I have some private matters in past half year.
So, we can first start to draw some 1-D plot of data.
Thanks for calling me.

**Reply by kunxianhuang** [reply]
Wait, I remember that modmesh shows graphical plots. So, I should get familiar with that. 
We can start from each class of functions and limits of parameters.

**Reply by yungyuc** [reply]
@j8xixo12 improved the plotting a lot and we are planning for further [revamping of the GUI for releasing](https://github.com/orgs/solvcon/projects/8).  @kunxianhuang, could you please try to build the latest modmesh and play with the Euler solver GUI?
